### PR TITLE
Online Pause RD

### PR DESCRIPTION
**Pull request by voduyemi**
WI: [W-11565278](https://gus.lightning.force.com/a07EE000013ZXftYAG) Added support for pausing Elevate commitments

# Critical Changes

# Changes
- We now support pausing Elevate commitments.

# Issues Closed

# Community Ideas Delivered

# Features Intended for Future Release

# Features for Elevate Customers

# New Metadata

# Deleted Metadata

### CODE REVIEW

File: force-app/main/default/classes/PS_Request.cls [hunk 1]
```diff
@@ -258,7 +260,13 @@ public inherited sharing class PS_Request {
                 if (String.isNotBlank(commitmentId)) {
                     value += '/' + commitmentId;
                 }
-
+            } else if (endpoint == ElevateEndpoint.COMMITMENT_PAUSE) {
+                value = String.format(
+                    ENDPOINT_COMMITMENT_CANCEL,
```
Comment (lparrott) [inline force-app/main/default/classes/PS_Request.cls:265]: Should this be `ENDPOINT_COMMITMENT_PAUSE`?

File: force-app/main/default/classes/RD2_PauseForm_TEST.cls [hunk 2]
```diff
@@ -185,6 +185,48 @@ public with sharing class RD2_PauseForm_TEST {
...
+    private static void mockRecordCreatedCalloutResponse() {
+        UTIL_Http_TEST.mockRecordCreatedCalloutResponse(mockSuccessResponseBody());
+    }
+
+
+    private static String mockSuccessResponseBody() {
+        return '{' +
+            '"startTimestamp":"' + System.today()  +'",' +
+            '"endTimestamp":"' + System.today().addDays(5) + '",' +
+            '"reason":"' + PAUSED_REASON_VALUE +
```
Comment (lparrott) [inline force-app/main/default/classes/RD2_PauseForm_TEST.cls:226]: Should we test with `reasonComment` or will that never be used?

File: force-app/main/default/classes/RD2_RecurringDonation.cls [hunk 3]
```diff
@@ -89,6 +89,10 @@ public inherited sharing class RD2_RecurringDonation {
         this(rd, rd?.npe03__Donations__r);
     }
 
+    public npe03__Recurring_Donation__c getRecord (){
+        return rd;
```
Comment (lparrott) [inline force-app/main/default/classes/RD2_RecurringDonation.cls:93]: This already exists as `getSObject`

**Review by lparrott — CHANGES_REQUESTED** [review]
Looking good, I just had a few questions/comments.

File: force-app/main/default/classes/RD2_PauseForm_TEST.cls (hunk 2 above)
Comment (voduyemi) [reply force-app/main/default/classes/RD2_PauseForm_TEST.cls:226]: It seems it will never be used I had asked about it from @force2b

File: force-app/main/default/classes/RD2_RecurringDonation.cls (hunk 3 above)
Comment (voduyemi) [reply force-app/main/default/classes/RD2_RecurringDonation.cls:93]: Updated!

File: force-app/main/default/classes/PS_Request.cls (hunk 1 above)
Comment (voduyemi) [reply force-app/main/default/classes/PS_Request.cls:265]: Updated!

**Review by lparrott — APPROVED** [review]

File: force-app/main/default/classes/RD2_CommitmentService.cls [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+    * @description Logs an error record on an commitment error response
+    * @param rd Recurring Donation record
+    * @param response Payments API response
+    */
+    private void processResponse(npe03__Recurring_Donation__c rd, UTIL_Http.Response response) {
+        if (isCommitmentSuccess(response)) {
+            return;
+        }
+
+        Id recordId = rd.Id != null
```
Comment (npsp-reedestockton) [inline force-app/main/default/classes/RD2_CommitmentService.cls:164]: From this point forward can be abstracted out into a separate method.  It is identical to the version below.

**Review by npsp-reedestockton — COMMENTED** [review]
Looks good!  Just one small request.

File: force-app/main/default/classes/RD2_CommitmentService.cls (hunk 4 above)
Comment (voduyemi) [reply force-app/main/default/classes/RD2_CommitmentService.cls:168]: Updated!

**Review by npsp-reedestockton — APPROVED** [review]
Looks great!

File: force-app/main/default/classes/RD2_CommitmentService.cls [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+    public UTIL_Http.Response handleCommitmentPause(RecurringDonationSchedule__c schedule, RD2_RecurringDonation rd) {
+        UTIL_Http.Response response;
+
+        PS_CommitmentRequest.RequestBody requestBody = new PS_CommitmentRequest().getPauseRequestBody(schedule);
+
+        response = sendRequest(rd.getSObject()?.CommitmentId__c, JSON.serialize(requestBody),
+            PS_Request.ElevateEndpoint.COMMITMENT_PAUSE);
+
+        System.debug(JSON.serializePretty(response));
```
Comment (npsp-reedestockton) [inline force-app/main/default/classes/RD2_CommitmentService.cls:66]: This should be removed.

File: force-app/main/default/classes/RD2_PauseForm_TEST.cls [hunk 6]
```diff
@@ -185,6 +185,48 @@ public with sharing class RD2_PauseForm_TEST {
...
+        pause.pausedReason = new RD2_PauseForm_CTRL.PausedReason();
+        pause.pausedReason.value = 'Unknown';
+
+        RD2_PauseForm_CTRL.savePause(JSON.serialize(pause));
+
+        Test.stopTest();
+
+        RecurringDonationSchedule__c schedule = [SELECT Id FROM RecurringDonationSchedule__c][0];
+
+        System.assertNotEquals(null, schedule, 'The Commitment response should be returned');
```
Comment (npsp-reedestockton) [inline force-app/main/default/classes/RD2_PauseForm_TEST.cls:214]: Maybe add a new assert that tests response is not null.  The message on this one does not match what is tested.
